torch/optim/svrg.py

```diff
@@ -0,0 +1,165 @@
+        dampening (float, optional): dampening for momentum (default: 0)
+        nesterov (bool, optional): enables Nesterov momentum (default: False)
+
+    .. _Accelerating stochastic gradient descent using predictive variance reduction:
+    https://papers.nips.cc/paper/4937-accelerating-stochastic-gradient-descent-using-predictive-variance-reduction.pdf
+    """
+    def __init__(self, params, snapshot_params=None, lr=required, momentum=0, dampening=0,
+                 weight_decay=0, nesterov=False):
+        if lr is not required and lr < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if momentum < 0.0:
+            raise ValueError("Invalid momentum value: {}".format(momentum))
+        if weight_decay < 0.0:
+            raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+
+        defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
+                        weight_decay=weight_decay, nesterov=nesterov)
+        super(SVRG, self).__init__(params, defaults)
+
+        # Making param groups out of the snapshot parameters
+        snapshot_param_groups = list(snapshot_params)
+        if len(snapshot_param_groups) == 0:
+            raise ValueError("optimizer got an empty snapshot parameter list")
+        if not isinstance(snapshot_param_groups[0], dict):
+            snapshot_param_groups = [{'snapshot_params': snapshot_param_groups}]
+
+        self.add_snapshot_param_group(snapshot_param_groups)
+
+    def step(self, closure=None):
+        r"""Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss, snapshot_loss = None, None
+        if closure is not None:
+            loss = closure()
+            if isinstance(loss, tuple):
+                loss, snapshot_loss = loss
+
+        for group in self.param_groups:
+            weight_decay = group['weight_decay']
+            momentum = group['momentum']
+            dampening = group['dampening']
+            nesterov = group['nesterov']
+            for idx, p in enumerate(group['params']):
+                if p.grad is None:
+                    continue
+                state = self.state[p]
+                # State initialization, add the average gradient to the state
+                if len(state) == 0:
+                    state['average_gradient'] = torch.zeros_like(p.grad.data)
+                average_gradient = state['average_gradient']
+                snapshot_params = group['snapshot_params'][idx]
+                # gradient data
+                d_p = p.grad.data
+                # subtract the average gradient
+                d_p.add_(-1, average_gradient)
```

Comment:
Please correct me if I'm wrong, but shouldn't we just add the average_gradient rather than taking the negative of it, since in the end you multiply `d_p` by `-lr`?

Reply:
@mengxixi You're right again. It seems I flipped the sign for the average gradient and the snapshot's gradient. I will fix it too

Reply:
hey, Please correct me if I'm wrong, at the end of d_p it becomes d_p = snapshot_gradient + gradient_data - average gradient  and for the weights p=p-lr*d_p. 
but on the other hand in the algorithm article we find d_p = snapshot_gradient - gradient_data + average = snapshot_gradient - (gradient_data - average). So is there a problem or no ?
![Capture](https://user-images.githubusercontent.com/54741713/83927285-c1fc0300-a78c-11ea-877f-56706cd8177f.PNG)

Reply:
```
# In the article the equation is: W_t = W_{t-1} - lr * ( gradient_{t-1} - snapshot_gradient + average )
# d_p = gradient_data
d_p = p.grad.data
# d_p = gradient_data + average_gradient
d_p.add_(average_gradient)
# d_p  = gradient_data + average_gradient - snapshot_gradient
# Which is the same as
d_p.add_(-1, snapshot_params.grad.data)
...
# p = p - lr * d_p
p.data.add_(-group['lr'], d_p)
```
I don't see the issue here. It is the same as in the article, no?

Reply:
yes it's the same thing, the problem is that you didn't change your script in the svrg.py file
              # gradient data
                d_p = p.grad.data
                # subtract the average gradient
                d_p.add_(-1, average_gradient)
                # add the snapshot gradient
                if snapshot_params.grad is not None:
                    d_p.add_(snapshot_params.grad.data)

Reply:
This is an outdated change you're seeing here. Please see the latest commit, it was fixed there.

Reply:
ok, thank you